doc/lspconfig.txt

```diff
@@ -1,242 +1,659 @@
+
+- {on_attach} `function(client, bufnr)`
+
+  Function called by neovim's built-in client after attaching a buffer to a
+  language server. Allows setting options (local or otherwise) conditioned
+  after a language server attaches, or overriding client options. Most
+  commonly used for settings buffer local keybindings.
 
+- {root_dir} `function(filename, bufnr)`
+
+  see |lspconfig-root-dir|
+
+- {settings} `table <string, string|table|bool>`  
+
+  The `settings` table is sent in `on_init` via a
+  `workspace/didChangeConfiguration` notification from the neovim client to
+  the language server. These settings allow a user to change optional runtime
+  settings of the language server. 
+
+  The lspconfig document generation pipeline converts the `package.json`
+  distributed by many language servers that includes optional settings into
+  documentation in `servers.md`. As a worked example, to send the following:
+
+  `pyright.disableLanguageServices`**: `boolean`
+  `pyright.disableOrganizeImports`**: `boolean`
+
+  The nested keys need to be translated into a nested table and passed to
+  settings as follows:
 >
-    lua << EOF    
-    vim.lsp.set_log_level("debug")    
-    EOF
+  require('lspconfig').pyright.setup{
+    settings = {
+      pyright = {
+          disableLanguageServices = true,
+          disableOrganizeImports  = true,
+        }
+    }
+  }
 <
+  
+  Note, the autogenerated settings occasionally include vscode specific
+  settings. If a setting is not respected by a language server, consult
+  upstream docuemntation.
 
+==============================================================================
+OVERRIDING GLOBAL DEFAULTS                         *lspconfig-override-defaults*
 
-Attempt to run the language server, and open the log with:
+The global defaults for all servers can be overridden by extending the
+default_config table. For a worked example, see below:
 
 >
-    :lua vim.cmd('e'..vim.lsp.get_log_path())
+  local lspconfig = require'lspconfig'
+  lspconfig.util.default_config = vim.tbl_extend(
+    "force",
+    lspconfig.util.default_config,
+    {
+      autostart = false,
+      handlers = {
+        ["window/logMessage"] = function(err, method, params, client_id)
+            if params and params.type <= vim.lsp.protocol.MessageType.Log then
+              vim.lsp.handlers["window/logMessage"](err, method, params, client_id)
+            end
+          end;
+        ["window/showMessage"] = function(err, method, params, client_id)
+            if params and params.type <= vim.lsp.protocol.MessageType.Warning.Error then
+              vim.lsp.handlers["window/showMessage"](err, method, params, client_id)
+            end
+          end;
+      }
+    }
+  )
+<
+`setup {}` can additionally override these defaults in subsequent calls.
+
+==============================================================================
+SERVER CONFIGURATIONS                                  *lspconfig-configuration*
+
```

Comment:
Shouldn't this be in the `setup` section?

Reply:
No, setup and the `config` interface are distinct